Sources/AsyncHTTPClient/HTTPDecompressor.swift

```diff
@@ -0,0 +1,151 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Swift Server Working Group and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import CNIOExtrasZlib
+import NIO
+import NIOHTTP1
+
+private enum CompressionAlgorithm: String {
+    case gzip
+    case deflate
+}
+
+extension z_stream {
+    public mutating func inflatePart(input: inout ByteBuffer, allocator: ByteBufferAllocator, consumer: (ByteBuffer) -> Void) {
+        input.readWithUnsafeMutableReadableBytes { dataPtr in
+            let typedPtr = dataPtr.baseAddress!.assumingMemoryBound(to: UInt8.self)
+            let typedDataPtr = UnsafeMutableBufferPointer(start: typedPtr, count: dataPtr.count)
+
+            self.avail_in = UInt32(typedDataPtr.count)
+            self.next_in = typedDataPtr.baseAddress!
+
+            repeat {
+                var buffer = allocator.buffer(capacity: 16384)
+                self.inflatePart(to: &buffer)
+                consumer(buffer)
```

Comment:
It's almost always better to return a value here rather than make an outcall, as you're running a risk that you'll crash in the face of an exclusive access violation (you hold `self` mutably here, and a recursive call into this object will re-enter and crash).

Given that the `input` is passed `inout`, it may be better to document that this API will not necessarily consume all the data from `input` and may need to be called in a loop. Indeed, having the caller provide the output buffer may also be useful.

Reply:
re-factored